modules/engine-prototype/src/main/scala/engine/Step.scala

```diff
@@ -0,0 +1,251 @@
+    object Ongoing {
+      final case class Standard(core: Core, tcsConfigured: Boolean, instConfigured: Boolean) extends Ongoing
+    }
+
+    sealed trait Failed extends F2
+    object Failed {
+      final case class Standard(core: Core, tcsFailed: Boolean, instFailed: Boolean, observationFailed: Boolean) extends F2
+      // TODO: final case class Standard(core: Core, tcsFailure: Option[Failure], instFailure: Option[Failure]) extends F2
+    }
+
+    sealed trait Current extends F2 {
+
+      // Accessor to the core Step
+      val core: Core = ???
+
+    }
+
+    object Current {
+      // Only a Pending Step in Current can be executed
+      sealed trait Pending extends Current {
+
+        def execute[F[_]](m: Sequence.State.Mutable[F])(implicit F: Effect[F], ec: ExecutionContext): F[Unit]
+
+      }
+
+      object Pending {
+
+        case class Standard(step: F2.Pending.Standard) extends Pending {
+
+          def execute[F[_]](m: Sequence.State.Mutable[F])(implicit F: Effect[F], ec: ExecutionContext): F[Unit] = {
+
+            def configureTCS: F[Boolean] =
+              Execution.configureTCS.flatMap {
+                case Right(_) =>
+                  m.modify(
+                    Sequence.State.current.modify {
+                      case Pending.Standard(s) => Ongoing.Standard(
+                        F2.Ongoing.Standard(s.core, tcsConfigured = true, instConfigured = false)
+                      )
+                      case Ongoing.Standard(s) => Ongoing.Standard(
+                        F2.Ongoing.Standard(s.core, tcsConfigured = true, instConfigured = true)
+                      )
+                      case Failed.Standard(s) => Failed.Standard(
+                        F2.Failed.Standard(
+                          s.core,
+                          tcsFailed = false,
+                          instFailed = true,
+                          observationFailed = true
+                        )
+                      ) // Instrument already failed
+                    }
+                  ) *> true.pure[F]
+
+                case Left(_) =>
+                  m.modify(
+                    Sequence.State.current.modify {
+                      case Failed.Standard(s) => Failed.Standard(
+                        F2.Failed.Standard(
+                          s.core,
+                          tcsFailed = true,
+                          instFailed = true,
+                          observationFailed = false
+                        )
+                      )
+                      // Instrument didn't fail yet
+                      case s => Failed.Standard(
+                        F2.Failed.Standard(
+                          s.core,
+                          tcsFailed = true,
+                          instFailed = false,
+                          observationFailed = false
+                        )
+                      )
+                    }
+                  ) *> false.pure[F]
+
+              }
```

Comment:
This is very verbose because I was too lazy to factor out the state modifications/observations into *Mutable* (I think I did this before that) and didn't create the lens for changing the parameters. But this is essentially what definining an `Action` should look like now.

The returned boolean is used to know if it failed or not to facilitate the continuation, but in reality a result with more info could be used instead.